### PR TITLE
Implement VK ID OAuth2 provider

### PR DESCRIPTION
**Pull request by hscokies**
Implement [VK ID](https://id.vk.com/about/business/go/docs/en/vkid/latest/vk-id/connection/start-integration/auth-without-sdk-web) OAuth2.

### CODE REVIEW

File: src/AspNet.Security.OAuth.VkId/VkIdAuthenticationErrors.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.VkId/VkIdAuthenticationErrors.cs]: There's no need to expose these as constants - they can just be specified in the code where they're used.

File: src/AspNet.Security.OAuth.VkId/VkIdAuthenticationExtensions.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.VkId/VkIdAuthenticationExtensions.cs]: Should all these references to "Vkontakte" be "VK ID"?

File: src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs]: There's no point adding new members that are immediately obsolete.

File: src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs]:
```suggestion
        var data  = RandomNumberGenerator.GetBytes(32);
```

File: src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs]: `["foo"] = "bar"` style preferred for new code please.

File: src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs]:
```suggestion
                var expiresAt = TimeProvider.GetUtcNow().AddSeconds(expiresIn);
```

File: src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs]:
```suggestion
        if (!context.Properties.Items.TryGetValue(VkIdAuthenticationConstants.AuthenticationProperties.DeviceId, out var deviceId) ||
```

File: src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs]:
```suggestion
        var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync(Context.RequestAborted));
```

File: src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs]: +1 on style here, also one key assignment per line.

File: src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs]:
```suggestion
        var content = await response.Content.ReadAsStringAsync(Context.RequestAborted);
```

File: src/AspNet.Security.OAuth.VkId/VkIdAuthenticationScopes.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.VkId/VkIdAuthenticationScopes.cs]:
```suggestion
/// List of <see href="https://id.vk.com/about/business/go/docs/vkid/latest/vk-id/connection/work-with-user-info/scopes">scopes</see>.
```

File: src/AspNet.Security.OAuth.VkId/AspNet.Security.OAuth.VkId.csproj
Comment (martincostello) [inline src/AspNet.Security.OAuth.VkId/AspNet.Security.OAuth.VkId.csproj:5]:
```suggestion
    <PackageValidationBaselineVersion>8.3.0</PackageValidationBaselineVersion>
    <TargetFrameworks>$(DefaultNetCoreTargetFramework)</TargetFrameworks>
```

**Review by martincostello — COMMENTED** [review]
Provider needs adding to README.

File: src/AspNet.Security.OAuth.VkId/VkIdAuthenticationExtensions.cs
Comment (hscokies) [reply src/AspNet.Security.OAuth.VkId/VkIdAuthenticationExtensions.cs]: Not sure 😅
VK ID = OAuth2.1 vkontakte implementation

File: test/AspNet.Security.OAuth.Providers.Tests/VkId/VkIdTests.cs
Comment (martincostello) [inline test/AspNet.Security.OAuth.Providers.Tests/VkId/VkIdTests.cs]:
```suggestion
```

**Review by martincostello — COMMENTED** [review]
(I see the PR changes, but I'm on vacation until next Wednesday so my ability to properly review/merge will be sporadic until then)

File: README.md
Comment (martincostello) [inline README.md]: Please revert all these formatting changes and just add the new line at the appropriate point in the A-Z order. Otherwise it's just ongoing maintenance to adjust for alignment as new providers are added.

File: src/AspNet.Security.OAuth.VkId/VkIdAuthenticationDefaults.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.VkId/VkIdAuthenticationDefaults.cs]: These should be `static readonly` like in our other providers so that if they ever need to change (there is past precedent for other providers), the values aren't "burned in" to any call sites:

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/82d20e3a415be2203257ecc1222edde5766961bb/src/AspNet.Security.OAuth.Amazon/AmazonAuthenticationDefaults.cs#L34-L47

File: src/AspNet.Security.OAuth.VkId/VkIdAuthenticationScopes.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.VkId/VkIdAuthenticationScopes.cs]: If we make it less specific, then if new fields are added to their API our documentation isn't wrong.

```suggestion
    /// Grants access to personal information.
```

File: src/AspNet.Security.OAuth.VkId/VkIdAuthenticationExtensions.cs
Comment (martincostello) [reply src/AspNet.Security.OAuth.VkId/VkIdAuthenticationExtensions.cs]: The reason I ask is because it's referred to as "VK ID" elsewhere in the XML documentation comments.

File: src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs]: I can't find the comment where it was originally stated, but please use a non-primary constructor for consistency with the rest of the codebase (adopting them was suggested once, but rejected).

File: src/AspNet.Security.OAuth.VkId/VkIdAuthenticationExtensions.cs
Comment (hscokies) [reply src/AspNet.Security.OAuth.VkId/VkIdAuthenticationExtensions.cs]: Yea, changed it to VK ID to not confuse people

File: src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs]: To match the other implementations.

```suggestion
        IOptionsMonitor<VkIdAuthenticationOptions> options,
        ILoggerFactory logger,
        UrlEncoder encoder)
```
```suggestion
        [NotNull] IOptionsMonitor<VkIdAuthenticationOptions> options,
        [NotNull] ILoggerFactory logger,
        [NotNull] UrlEncoder encoder)
```

File: src/AspNet.Security.OAuth.VkId/VkIdAuthenticationOptions.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.VkId/VkIdAuthenticationOptions.cs:1]: Looking at the documentation, PKCE is supported, so `UsePkce` should be defaulted to `true` and the appropriate changes made to `ExchangeCodeAsync()` in the handler (search PKCE in the existing handlers to see the sort of changes needed).

Comment (hscokies) [reply src/AspNet.Security.OAuth.VkId/VkIdAuthenticationOptions.cs:1]: Looks like it's mandatory to use PKCE
image
image
Should I still specify UsePkce default value?

Comment (martincostello) [reply src/AspNet.Security.OAuth.VkId/VkIdAuthenticationOptions.cs:1]: Yes, as the base class doesn't enable it by default.

Comment (hscokies) [reply src/AspNet.Security.OAuth.VkId/VkIdAuthenticationOptions.cs:1]: Added UsePkce default value, AuthenticationProperties check already at [L166](https://github.com/HScokies/AspNet.Security.OAuth.Providers.Vkid/blob/7c69c627aaec7d5ae251471861cbae4119883dfb/src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs#L166-L166)

File: src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs]: Can you factor this to match the pattern used in the other providers similar to this please? For example, the error and success cases are other way around and the HTTP status code isn't being explicitly checked, so a JSON response without an `error` property is always treated as success, even if it returns a non-success response.

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/82d20e3a415be2203257ecc1222edde5766961bb/src/AspNet.Security.OAuth.Deezer/DeezerAuthenticationHandler.cs#L47-L59

Similarly for other methods in this handler class in terms of layout and style - that makes the providers easier to maintain if we have to make a similar change in a future version to all 90+ providers.

Not being similar is partly why I didn't realise this already handled PKCE at a glance, because it was structured differently and the accompanying comment like you find in the other providers.

Comment (hscokies) [reply src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs]: Got it! gonna refactor it later this week.
Also, should I add logger to this handler? If yes, are there any docs about event IDs?

Comment (martincostello) [reply src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs]: Just whatever logging makes sense in the same style as the other providers. The event IDs are just arbitrary numbers starting from 1 as the provider is implemented.

File: src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs]:
```suggestion
        // OAuth2 10.12 CSRF
        if (!ValidateCorrelationId(properties))
        {
            return HandleRequestResult.Fail("Correlation failed.", properties);
        }
```

File: src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs]:
```suggestion
            return HandleRequestResult.Fail("Code was not found.", properties);
```

File: src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs]:
```suggestion
            return HandleRequestResult.Fail("Device ID was not found.", properties);
```

File: src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs
Comment (martincostello) [inline src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs]: Could you invert this so the failure case is in the `if` and the success case is the end of the method please?

File: test/AspNet.Security.OAuth.Providers.Tests/VkId/VkIdTests.cs
Comment (martincostello) [inline test/AspNet.Security.OAuth.Providers.Tests/VkId/VkIdTests.cs]: Do something like this, then you only need to set one extra property:

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/34f683a19d855fdc133dbd176cc0ca9766f422b0/test/AspNet.Security.OAuth.Providers.Tests/Okta/OktaTests.cs#L15-L19

Why is it that you need to turn this off?

Comment (hscokies) [reply test/AspNet.Security.OAuth.Providers.Tests/VkId/VkIdTests.cs]: I'm out of ideas how to mock IDataProtector to pass tests failing on this [line](https://github.com/HScokies/AspNet.Security.OAuth.Providers.Vkid/blob/366b9092112d7a08a4ca28f4cf39921973d99d63/src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs#L199) :(

Comment (martincostello) [reply test/AspNet.Security.OAuth.Providers.Tests/VkId/VkIdTests.cs]: You'll need to configure the HTTP request interceptor to echo back the state in the response body in some way. For all the other providers this is handled [here](https://github.com/HScokies/AspNet.Security.OAuth.Providers.Vkid/blob/366b9092112d7a08a4ca28f4cf39921973d99d63/test/AspNet.Security.OAuth.Providers.Tests/Infrastructure/LoopbackRedirectHandler.cs#L83-L121) by adjusting the callback URL.

We shouldn't really be building in an option there's no real reason to ever turn off from a consumer's point of view _just_ to get the tests to pass.

Comment (hscokies) [reply test/AspNet.Security.OAuth.Providers.Tests/VkId/VkIdTests.cs]: phew, looks like I've finally got it :D

File: test/AspNet.Security.OAuth.Providers.Tests/VkId/VkIdTests.cs
Comment (martincostello) [inline test/AspNet.Security.OAuth.Providers.Tests/VkId/VkIdTests.cs]: Could you make this use a slightly more realistic value that matches a non-empty `scope` value in the payload?

**Review by martincostello — APPROVED** [review]
Just one minor comment regarding the tests.

File: src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs
Comment (kevinchalet) [inline src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs]: It's not technically wrong, but `!ValidateCorrelationId(properties)` would be easier to read (and consistent with the other providers).

File: src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs
Comment (kevinchalet) [inline src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs]: `state` in token responses is not really a standard thing and it's not clear what security benefits validating the `state` here provides. I'd just remove that part to simplify things.

File: src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs
Comment (kevinchalet) [inline src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs]: Looks like these temp variables aren't very useful: feel free to remove them and do the formatting directly in the dictionary initializer to save a few lines.

**Review by kevinchalet — COMMENTED** [review]
Thanks for your PR!

**Comment by kevinchalet** [comment]
(I find it funny that VK claims they are implementing "OAuth 2.1" - that hasn't been officially adopted yet - when all they're doing is implementing yet another non-standard version of OAuth with non-standard concepts like HTTP 200 errors and `device_id` 😄)

**Comment by hscokies** [comment]
yea, VK is a funny company...
image

**Review by kevinchalet — APPROVED** [review]
Looks good, thanks a lot for your contribution! 👍🏻 

I'll let @martincostello merge it in case he'd have additional feedback 😃

**Comment by martincostello** [comment]
Before merging this, can you confirm you've tested the provider against the production VK ID service and that you were able to successfully authenticate?

**Comment by martincostello** [comment]
Oh, and could you update this to `3` and `0` please?

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/8bff342db12c404ae6604158c7b5be63cec23471/eng/Versions.props#L5-L6

**Comment by hscokies** [comment]
> Before merging this, can you confirm you've tested the provider against the production VK ID service and that you were able to successfully authenticate?

Everything seems to work fine, yes
image
image

**Comment by martincostello** [comment]
Thanks again for your contribution - the new provider 📦 is now available from NuGet.org: https://www.nuget.org/packages/AspNet.Security.OAuth.VkId/8.3.0

**Pull request merged**
